# 01. Import libraries and data

In [2]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [3]:
# Define path
path = r'/home/scruffy/anaconda_projects/Instacart Basket Analysis/'

In [4]:
# Import dataframes
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'))
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'))
df_depts = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'departments.csv'))

In [10]:
print(df_ords.shape)
print(df_prods.shape)

(3421083, 7)
(49693, 5)


# 02. Arrange data

In [85]:
# Change data type for order_id
df_ords['order_id'] = df_ords['order_id'].astype('str')

# Transpose departments.csv data
df_depts_t = df_depts.T
new_header = df_depts_t.iloc[0]
df_depts_t_new = df_depts_t[1:]
df_depts_t_new.columns = new_header
print(df_depts_t_new)

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing


# 03. Set data dictionary

In [87]:
data_dict = df_depts_t_new.to_dict('index')

# 04. Exercise Step 2

In [89]:
# Change data type for user_id
df_ords['user_id'] = df_ords['user_id'].astype('str')
print(df_ords['user_id'].dtype)

object


# 05. Exercise Step 3

In [91]:
# Rename order_number without overwriting dataframe
df_ords.rename(columns = {'order_number':'number_of_orders_by_user'}, inplace = False)

order_id user_id eval_set  number_of_orders_by_user  order_dow  \
0        2539329       1    prior                         1          2   
1        2398795       1    prior                         2          3   
2         473747       1    prior                         3          3   
3        2254736       1    prior                         4          4   
4         431534       1    prior                         5          4   
...          ...     ...      ...                       ...        ...   
3421078  2266710  206209    prior                        10          5   
3421079  1854736  206209    prior                        11          4   
3421080   626363  206209    prior                        12          1   
3421081  2977660  206209    prior                        13          1   
3421082   272231  206209    train                        14          6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

# 06. Exercise Step 4

In [93]:
# Count values order_hour_of_day
df_ords_sorted = df_ords.value_counts('order_hour_of_day')
print(df_ords_sorted.iloc[:1])

order_hour_of_day
10    288418
Name: count, dtype: int64


## The busiest hour is 10:00 AM - 11:00 AM with 288,148 orders.

# 07. Exercise Step 5

In [96]:
# Look up department 4 in data dictionary
print(data_dict.get('4'))

{'department': 'produce'}


# 08. Exercise Step 6

In [98]:
# Subset to include only breakfast items
df_breakfast = df_prods[df_prods['department_id']==14]
print(df_breakfast.head())

     product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6  


# 09. Exercise Step 7

In [100]:
# Subset to include only dinner party items
df_party = df_prods[df_prods['department_id'].isin([5,7,12,20])]

# 10. Exercise Step 8

In [102]:
# Count rows and columns of resulting dataframe
print(df_party.shape)

(7650, 5)


## There are 7650 rows in the dinner party dataframe.

# 11. Exercise Step 9

In [105]:
# Subset orders.csv to user 1
df_user_1 = df_ords[df_ords['user_id']=='1']

# 12. Exercise Step 10

In [107]:
# Print descriptive stats for user 1
print(df_user_1.describe())

       order_number  order_dow  order_hour_of_day  days_since_prior_order
count     11.000000  11.000000          11.000000               10.000000
mean       6.000000   2.636364          10.090909               19.000000
std        3.316625   1.286291           3.477198                9.030811
min        1.000000   1.000000           7.000000                0.000000
25%        3.500000   1.500000           7.500000               14.250000
50%        6.000000   3.000000           8.000000               19.500000
75%        8.500000   4.000000          13.000000               26.250000
max       11.000000   4.000000          16.000000               30.000000


# 13. Exercise Step 12

In [109]:
# Export df_ords
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'))

# 14. Exercise Step 13

In [111]:
# Export df_depts_t_new
df_depts_t_new.to_csv(os.path.join(path, '02 Data','Prepared Data', 'departments_wrangled.csv'))